File: backend/src/domain/services/ActiveCartMerger.ts
import type { ActiveCartItem } from "@domain/entities/ActiveCart";

export class ActiveCartMerger {
	static merge(
		existingItems: readonly ActiveCartItem[],
		incomingItems: readonly ActiveCartItem[],
	): ActiveCartItem[] {
		const resultMap = new Map<string, ActiveCartItem>();

		for (const item of existingItems) {
			const key = ActiveCartMerger.compositeKey(item);
			resultMap.set(key, { ...item });
		}

		for (const item of incomingItems) {
			const key = ActiveCartMerger.compositeKey(item);
			resultMap.set(key, { ...item });
		}

		return Array.from(resultMap.values());
	}

	private static compositeKey(item: ActiveCartItem): string {
		return `${item.supermarket}:${item.productName}`;
	}
}

File: backend-dotnet/src/api/Domain/Services/ActiveCartMerger.cs
using SupermarketPlanner.Api.Domain.Entities;

namespace SupermarketPlanner.Api.Domain.Services;

public sealed class ActiveCartMerger
{
    public static IReadOnlyList<ActiveCartItem> Merge(
        IReadOnlyList<ActiveCartItem> existingItems,
        IReadOnlyList<ActiveCartItem> incomingItems)
    {
        var resultMap = new Dictionary<string, ActiveCartItem>();

        foreach (var item in existingItems)
            resultMap[CompositeKey(item)] = item;

        foreach (var item in incomingItems)
            resultMap[CompositeKey(item)] = item;

        return resultMap.Values.ToList();
    }

    private static string CompositeKey(ActiveCartItem item) =>
        $"{item.Supermarket}:{item.ProductName}";
}
